text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+## Table of Contents
+
+- [Motivation](#motivation)
+- [Protocol Details](#protocol-details)
+  - [Hello Handshake](#hello-handshake)
+  - [Export Request and Response](#export-request-and-response)
+  - [Client Modes](#client-modes)
+  - [Server Response](#server-response)
+  - [gRPC Transport](#grpc-transport)
+  - [Other Transports](#other-transports)
+- [Implementation Recommendations](#implementation-recommendations)
+  - [Pipelined Mode](#pipelined-mode)
+  - [Multi-Destination Exporting](#multi-destination-exporting)
+- [Trade-offs and Mitigations](#trade-offs-and-mitigations)
+- [Future Versions and Interoperability](#future-versions-and-interoperability)
+- [Prior Art, Alternatives and Future Possibilities](#prior-art-alternatives-and-future-possibilities)
+- [Open Questions](#open-questions)
+- [Appendix A - Protocol Buffer Definitions](#appendix-a---protocol-buffer-definitions)
+- [Appendix B - Performance Benchmarks](#appendix-b---performance-benchmarks)
+- [Glossary](#glossary)
+- [Acknowledgements](#acknowledgements)
+- [Author's Address](#authors-address)
+
+
+## Motivation
+
+OTLP is a general-purpose telemetry data delivery protocol designed in the scope of OpenTelemetry project. It is an incremental improvement of OpenCensus protocol. Compared to OpenCensus protocol OTLP has the following improvements:
+
+- Ensures high reliability of data delivery and clear visibility when the data cannot be delivered. OTLP uses acknowledgements and retry mechanisms to implement reliable delivery.
+
+- It is friendly to Level 7 Load Balancers and allows them to correctly map imbalanced incoming traffic to a balanced outgoing traffic. This allows to efficiently operate large networks of nodes where telemetry data generation rates change over time.
+
+- Allows backpressure signalling from telemetry data destinations to sources. This is important for implementing reliable multi-hop telemetry data delivery all the way from the source to the destination via intermediate nodes, each having different processing capacity and thus requiring different data transfer rates.
+
+## Protocol Details
+
+OTLP defines the encoding of telemetry data and the protocol used to exchange data between the client and the server.
+
+This specification defines how OTLP is implemented over [gRPC](https://grpc.io/) and specifies corresponding [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview) schema. Future extensions to OTLP may define implementations over other transports. For details of gRPC service definition see section [gRPC Transport](#grpc-transport).
+
+OTLP is a request/response style protocols: the clients send requests, the server replies with corresponding responses. There are 2 supported requests and response types: `Hello` and `Export`.
+
+### Hello Handshake
```

Comment:
`Hello` is used for client and server to discover each others capabilities. Strictly speaking it is not needed right now, because there are no optional capabilities defined in OTLP yet.

In the future when we define new capabilities they will be encoded in `Hello` message payload. The client will send a `Hello` request and expect a response from the server.

Now, instead of introducing the `Hello` request/response now we could suggest an alternate approach: introduce `Hello` when optional capabilities are designed.

The problem is we want old servers to be able to speak with new clients. When new client makes a `Hello` request but old server does not know what is `Hello` request it will return generic `Unimplemented` gRPC error. This requires the client to separate process this error case and assume that when this error is returned it means that the server is old. This makes `Hello` response processing on the client side non-uniform. Instead of just checking the response message client now has to special-case certain gRPC error codes.

To avoid this I suggest to introduce `Hello` request/response now. Servers are required to respond with empty `Hello` message when they receive the request. This makes future extensions of OTLP by new capabilities more uniform.

If you think this is unnecessary we can remove `Hello` from current spec, but I think it will make future extensions a bit less clean.

Reply:
`Hello` removed.